### PR TITLE
Add Bloodhound Community Edition

### PR DESCRIPTION
**Pull request by QU35T-code**
# Description

Added **Bloodhound Community Edition**: https://github.com/SpecterOps/BloodHound

**Commands :**

* `bloodhound-ce` (starts psql database and bloodhound service)
* `bloodhound-ce-stop` (stop the psql database and the bloohound service)
* `bloodhoud-ce-reset` (reset the psql database and stop the service + bloodhound service)

Here are the ports used :

`5432` -> PostgreSQL database
`1030` -> Bloodhound-CE web service
`7474 et 7687` -> neo4j database

On first launch the `admin` password will be indicated in the console :

```bash
{"level":"info","time":"2023-08-30T22:29:03.706604982Z","message":"# Initial Password Set To:    1NG9pUCJwQKuJDf1k1TzH63giy_tGPRQ    #"}
```

image

image

image

### CODE REVIEW

**Comment by QU35T-code** [comment]
I think there is a problem with the azure parser. I have to fix before merge

File: sources/install/package_base.sh [hunk 1]
```diff
@@ -244,6 +244,7 @@ function install_gf() {
 function post_install() {
     # Function used to clean up post-install files
     colorecho "Cleaning..."
+    apt update
```
Comment (ShutdownRepo) [inline sources/install/package_base.sh:247]: Why running an `apt update` here?
And wouldn't `apt-get update` be more suited for a script usage

File: sources/install/package_ad.sh [hunk 2]
```diff
@@ -125,6 +125,53 @@ function install_bloodhound() {
...
+    cd ./azurehound
+    wget \
+    https://github.com/BloodHoundAD/AzureHound/releases/download/v2.0.5/azurehound-linux-amd64.zip \
+    https://github.com/BloodHoundAD/AzureHound/releases/download/v2.0.5/azurehound-linux-arm64.zip
+    cd /opt/tools/BloodHound
+    cp ./dist/bhapi ./bloodhound
+    mkdir -p /etc/bloodhound/collectors/sharphound/
+    cp -r ./sharphound-v2.0.0.zip* /etc/bloodhound/collectors/sharphound/
+    mkdir -p /etc/bloodhound/collectors/azurehound/
+    cp -r ./azurehound/azurehound-linux-a* /etc/bloodhound/collectors/azurehound/
```
Comment (ShutdownRepo) [inline sources/install/package_ad.sh:163]: In my opinion, this should be handled by Exegol-resources

File: sources/install/package_ad.sh [hunk 3]
```diff
@@ -125,6 +125,53 @@ function install_bloodhound() {
...
+    cp ./dockerfiles/configs/bloodhound.config.json ./
+    sed -i "s#app-db#127.0.0.1##" bloodhound.config.json
+    sed -i "s#graph-db#127.0.0.1##" bloodhound.config.json
+    sed -i "s#8080#1030##" bloodhound.config.json
+    sed -i "s#neo4j:bloodhoundcommunityedition#neo4j:exegol4thewin##" bloodhound.config.json
+    cp /root/sources/assets/bloodhound-ce/*.sh /opt/tools/bin/
```
Comment (ShutdownRepo) [inline sources/install/package_ad.sh:169]: wouldn't symlinks be more suited?

File: sources/install/package_base.sh (hunk 1 above)
Comment (QU35T-code) [reply sources/install/package_base.sh:247]: Because when you launch a new container, apt is not up to date and you cannot install a packet without doing apt update. I haven't really thought about which one is better between apt update and apt-get update

File: sources/install/package_ad.sh (hunk 2 above)
Comment (QU35T-code) [reply sources/install/package_ad.sh:163]: Right ! We can install sharphound + azurehound from my_resources and link to /etc/bloodhound/collectors after

File: sources/install/package_ad.sh (hunk 3 above)
Comment (QU35T-code) [reply sources/install/package_ad.sh:169]: Bash binaries are in the exegol installation folder. Making a symlink will not work because after post_install the folder will no longer exist. I understood well ?

**Comment by QU35T-code** [comment]
I will also add a clean to the Bloodhound folder after the build to save space

File: sources/install/package_base.sh (hunk 1 above)
Comment (gbe) [reply sources/install/package_base.sh:247]: According to the [manual page](https://manpages.debian.org/buster/apt/apt.8.en.html), in scripts apt-get must be preferred, while apt is for interactive sessions.

**Comment by Dramelac** [comment]
Very nice ! 
I have a concern to share, because there is multiple local service running now, when using multiple container in host mode this might create some conflict, would it be possible to use random port for local service ? It may be manager by the start script to init / customize the config on first use

**Comment by QU35T-code** [comment]
> Very nice ! I have a concern to share, because there is multiple local service running now, when using multiple container in host mode this might create some conflict, would it be possible to use random port for local service ? It may be manager by the start script to init / customize the config on first use

Sure ! We can. All we need to edit it's the config file -> `/opt/tools/BloodHound/bloodhound.config.json`. I will add (later) a specific comment above so that the wrapper has the correct steps !

File: sources/install/package_base.sh (hunk 1 above)
Comment (QU35T-code) [reply sources/install/package_base.sh:247]: That's right !

Comment (ShutdownRepo) [reply sources/install/package_base.sh:247]: 100% agree with @gbe 
Let’s remove the apt update here

File: sources/install/package_ad.sh (hunk 2 above)
Comment (ShutdownRepo) [reply sources/install/package_ad.sh:163]: Not 100% necessary to link it there is it?
But yeah let’s have the install made by exegol-resources

File: sources/install/package_ad.sh (hunk 3 above)
Comment (ShutdownRepo) [reply sources/install/package_ad.sh:169]: My bad you’re right

File: sources/install/package_ad.sh (hunk 2 above)
Comment (QU35T-code) [reply sources/install/package_ad.sh:163]: We have to link because bloodhound needs these zip to work (when starts). Note : We can change the Path if necessary

**Comment by QU35T-code** [comment]
I did what was necessary to delete the build folder and indicate in the PR description how we can modify the port configuration (cc @Dramelac). I need to add the resources to the exegol-resources repo

**Comment by Dramelac** [comment]
A manually description is a first step but not really useful in real world with UX, nobody will change anything. If we want to do things right, you should try to add an automatically and transparent mechanism

**Comment by ShutdownRepo** [comment]
> A manually description is a first step but not really useful in real world with UX, nobody will change anything. If we want to do things right, you should try to add an automatically and transparent mechanism

agreed

**Comment by QU35T-code** [comment]
Don't know why but node hasn't the same version :

```bash
root@exegol-deploy /workspace # node --version
v20.5.1
```

```
2023-09-01T09:39:54.5488137Z #24 1004.5 [1;34m[EXEGOL] DEBUG : Node version[0m
2023-09-01T09:39:54.6998282Z #24 1004.5 v12.22.12
```

**Comment by ShutdownRepo** [comment]
Putting PR on hold, we'll merge #216 first, and then work from there

**Comment by ShutdownRepo** [comment]
#216 is now merged, removing the hold on this PR

File: sources/install/package_ad.sh [hunk 4]
```diff
@@ -118,13 +118,81 @@ function install_bloodhound() {
     colorecho "Installing BloodHound from sources"
     git -C /opt/tools/ clone --depth 1 https://github.com/BloodHoundAD/BloodHound/
     mv /opt/tools/BloodHound /opt/tools/BloodHound4
-    zsh -c "source ~/.zshrc && cd /opt/tools/BloodHound4 && nvm install 16.13.0 && nvm use 16.13.0 && npm install -g electron-packager && npm install && npm run build:linux"
+    zsh -c "source ~/.zshrc && cd /opt/tools/BloodHound4 && nvm install 16.13.0 && nvm use 16.13.0 && npm install -g electron-packager && npm install && npm run build:linux && nvm use default"
```
Comment (ShutdownRepo) [inline sources/install/package_ad.sh:121]: why the `nvm use default` at the end?

File: sources/install/package_ad.sh [hunk 5]
```diff
@@ -118,13 +118,81 @@ function install_bloodhound() {
...
 }
 
+function install_bloodhound-ce() {
+    # CODE-CHECK-WHITELIST=add-aliases,add-history
+    colorecho "Installing BloodHound-CE"
+
+    # Installing & Configuring the database
+    fapt postgresql postgresql-client
+    service postgresql start
+    cd /tmp
```
Comment (ShutdownRepo) [inline sources/install/package_ad.sh:135]: please add `|| exit` for all `cd` instructions

File: sources/install/package_ad.sh [hunk 6]
```diff
@@ -118,13 +118,81 @@ function install_bloodhound() {
...
+    sudo -u postgres psql -c "CREATE USER bloodhound WITH PASSWORD 'bloodhoundcommunityedition';"
+    sudo -u postgres psql -c "CREATE DATABASE bloodhound;"
+    sudo -u postgres psql -c "GRANT ALL PRIVILEGES ON DATABASE bloodhound TO bloodhound;"
```
Comment (ShutdownRepo) [inline sources/install/package_ad.sh:138]: `sudo` not needed

File: sources/install/package_ad.sh [hunk 7]
```diff
@@ -118,13 +118,81 @@ function install_bloodhound() {
...
+
+    # Move files
+    mkdir /opt/tools/BloodHound-CE/
+    mkdir -p /opt/tools/BloodHound-CE/collectors/sharphound/
+    mkdir -p /opt/tools/BloodHound-CE/collectors/azurehound/
+    mkdir /opt/tools/BloodHound-CE/work
+    cp ./dist/bhapi /opt/tools/BloodHound-CE/bloodhound
+    cp -r ./sharphound-v2.0.0.zip* /opt/tools/BloodHound-CE/collectors/sharphound/
+    cp -r ./azurehound/azurehound-linux-a* /opt/tools/BloodHound-CE/collectors/azurehound/
+    cp ./dockerfiles/configs/bloodhound.config.json /opt/tools/BloodHound-CE/
```
Comment (ShutdownRepo) [inline sources/install/package_ad.sh:176]: Please add `-v` flags

**Review by ShutdownRepo — CHANGES_REQUESTED** [review]

File: sources/install/package_ad.sh (hunk 6 above)
Comment (QU35T-code) [reply sources/install/package_ad.sh:138]: Need to be postgres user to perform this action

Comment (ShutdownRepo) [reply sources/install/package_ad.sh:138]: Indeed

File: sources/assets/bloodhound-ce/bloodhound-ce-reset [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+sudo -u postgres psql -c "DROP database bloodhound"
+sudo -u postgres psql -c "CREATE DATABASE bloodhound;"
```
[the remark this answers is not in the document]
Comment (ShutdownRepo) [reply sources/assets/bloodhound-ce/bloodhound-ce-reset:4]: https://github.com/ThePorgs/Exegol-images/pull/222#discussion_r1341273826

**Comment by QU35T-code** [comment]
I will make the changes at the beginning of the week

File: sources/install/package_ad.sh (hunk 4 above)
Comment (QU35T-code) [reply sources/install/package_ad.sh:121]: Because bh needs to build with node 16.13.0 and at the end of the installation, I switch to the default version of node (v20.8.0)

**Comment by ShutdownRepo** [comment]
Made some changes, it works when running the install function in a live container.
Starting a local image build in parallel to the pipeline checks
If I get green light on both ends, I will approve this PR for merge. However, given the "complexity" of the service and our in-ability to have a solid unit test, we won't be able to guarantee it works during the pipeline checks

**Comment by ShutdownRepo** [comment]
Pipeline is a success, local tests are a success 👀 
@Dramelac @QU35T-code final review please ?

File: sources/install/package_ad.sh [hunk 9]
```diff
@@ -771,6 +836,7 @@ function install_pcredz() {
 
 function install_pywsus() {
     colorecho "Installing pywsus"
+    fapt libxml2-dev libxslt-dev
```
Comment (QU35T-code) [inline sources/install/package_ad.sh:839]: This is not part of bloodhound

**Comment by QU35T-code** [comment]
Green lights !!

File: sources/assets/bloodhound-ce/bloodhound-ce [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+echo "##### EXEGOL SAYS ##### Listening on http://$(jq -r .bind_addr < /opt/tools/BloodHound-CE/bloodhound.config.json)"
+echo "#####             ##### Don't forget to run neo4j-start"
+echo "#####             ##### Login : admin"
+echo "#####             ##### Password : intial password will only be printed on the first run of bloodhound, it then needs to be changed in the Web UI"
+echo "#####             #####            if you lost the password, run bloodhound-ce-reset."
+
+if ! pg_isready -q; then
```
Comment (Dramelac) [inline sources/assets/bloodhound-ce/bloodhound-ce:9]: Isn't there any equivalent for neo4j while we are at it ?

File: sources/install/package_ad.sh [hunk 11]
```diff
@@ -140,6 +140,71 @@ function install_bloodhound() {
...
+    cd /opt/tools/BloodHound-CE/src/ || exit
+    catch_and_retry VERSION=v999.999.999 CHECKOUT_HASH="" python3 ./packages/python/beagle/main.py build --verbose --ci
+
+    # Collectors
+    mkdir -p /opt/tools/BloodHound-CE/collectors/sharphound
+    mkdir -p /opt/tools/BloodHound-CE/collectors/azurehound
+    ## SharpHound
+    local SHARPHOUND_URL
+    SHARPHOUND_URL=$(curl --location --silent "https://api.github.com/repos/BloodHoundAD/SharpHound/releases/latest" | grep 'SharpHound-.*.zip' | grep -v 'debug' | grep -o 'https://[^"]*')
+    wget --directory-prefix /opt/tools/BloodHound-CE/collectors/sharphound/ "$SHARPHOUND_URL"
```
Comment (Dramelac) [inline sources/install/package_ad.sh:173]: Sharphound shouldn't be part of exegol resources ?
However, we can add a symlink to the exegol resources sharphound directory ?

File: sources/install/package_ad.sh [hunk 12]
```diff
@@ -140,6 +140,71 @@ function install_bloodhound() {
...
+    local SHARPHOUND_NAME
+    SHARPHOUND_NAME=$(curl --location --silent "https://api.github.com/repos/BloodHoundAD/SharpHound/releases/latest" | grep -o 'SharpHound-.*.zip' | grep -v debug | uniq)
+    sha256sum "/opt/tools/BloodHound-CE/collectors/sharphound/$SHARPHOUND_NAME" > "/opt/tools/BloodHound-CE/collectors/sharphound/$SHARPHOUND_NAME.sha256"
+    ## AzureHound
+    local AZUREHOUND_URL_AMD64
+    local AZUREHOUND_URL_ARM64
+    AZUREHOUND_URL_AMD64=$(curl --location --silent "https://api.github.com/repos/BloodHoundAD/AzureHound/releases/latest" | grep 'azurehound-linux-arm64.zip' | grep -v 'sha' | grep -o 'https://[^"]*')
+    AZUREHOUND_URL_ARM64=$(curl --location --silent "https://api.github.com/repos/BloodHoundAD/AzureHound/releases/latest" | grep 'azurehound-linux-amd64.zip' | grep -v 'sha' | grep -o 'https://[^"]*')
+    wget --directory-prefix /opt/tools/BloodHound-CE/collectors/azurehound/ "$AZUREHOUND_URL_AMD64"
+    wget --directory-prefix /opt/tools/BloodHound-CE/collectors/azurehound/ "$AZUREHOUND_URL_ARM64"
```
Comment (Dramelac) [inline sources/install/package_ad.sh:183]: Same comment

**Review by Dramelac — APPROVED** [review]
Some comments and questions but seems good overall

File: sources/assets/bloodhound-ce/bloodhound-ce (hunk 10 above)
Comment (ShutdownRepo) [reply sources/assets/bloodhound-ce/bloodhound-ce:9]: We didn't implement this for bloodhound < bloodhound-ce, I wanted to keep the same behavior.
However, when the service and ports manager is implemented, I want to have neo4j managed when starting any version of bloodhound.
TL;DR: we can keep it that way for now, so that all versions of bloodhound behave the same. Will change later on with the srv manager utility being developped

File: sources/install/package_ad.sh (hunk 11 above)
Comment (ShutdownRepo) [reply sources/install/package_ad.sh:173]: In theory, yes.
In practice, bloodhound-ce requires the ingestors to be in a specific directory and checks that when starting. We'll probably need to have ingestors installed in the image as well as the resources

**Comment by ShutdownRepo** [comment]
Added a comment, following @Dramelac's review.
Skipping pipeline runs (they won't add anything), and merging 🥳
